{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "hairy-wireless",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from collections import Counter\n",
    "import numpy as np\n",
    "import json\n",
    "pd.options.mode.chained_assignment = None  # default='warn'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "headed-scott",
   "metadata": {},
   "source": [
    "### Table S3A: Combine CNV data from each strain into one dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "desirable-accent",
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_dict = json.load(open('../isolate_and_pop_info/pop_dict.json'))\n",
    "for strain, pops in pop_dict.items():\n",
    "    pop_dict[strain] = pops\n",
    "    \n",
    "strains = list(pop_dict.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "spare-premiere",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnv_files = !ls FREEC_output_250bp_window/*_CNVs\n",
    "\n",
    "df = pd.DataFrame()\n",
    "\n",
    "for i, f in enumerate(cnv_files):\n",
    "    \n",
    "    strain = f.replace('.CpSilv_core_only_bwa.sorted.bam_CNVs', '').replace('FREEC_output_250bp_window/','')\n",
    "    \n",
    "    if strain not in strains:\n",
    "        continue\n",
    "        \n",
    "    temp = pd.read_csv(f, sep='\\t', header=None,)\n",
    "    temp.columns=['chrom', 'start','stop', 'CN', 'type']\n",
    "    temp['strain'] = [strain for s in range(temp.shape[0])]\n",
    "\n",
    "    df = df.append(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "julian-utilization",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('../tables/TableS3A_allCNV.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "legitimate-fraction",
   "metadata": {},
   "source": [
    "### Combine CNV info across strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "moderate-thunder",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratio_files = !ls FREEC_output_250bp_window/*_ratio.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "respiratory-roman",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i,f in enumerate(ratio_files):\n",
    "    \n",
    "    strain = f.replace('FREEC_output_250bp_window/', '').replace('.CpSilv_core_only_bwa.sorted.bam_ratio.txt', '')  \n",
    "    if strain not in strains:\n",
    "        continue\n",
    "    \n",
    "    temp = pd.read_csv(f, sep='\\t')[['Chromosome','Start','CopyNumber']]\n",
    "    temp = temp.rename(columns={'CopyNumber':strain})\n",
    "    \n",
    "    if i == 0:\n",
    "        df = temp\n",
    "    else:\n",
    "        df = pd.merge(df,temp, on=['Chromosome','Start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "secure-tulsa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>2566_Venezuela</th>\n",
       "      <th>3796_Venezuela</th>\n",
       "      <th>4545-MICE_Venezuela</th>\n",
       "      <th>730334_Guatemala</th>\n",
       "      <th>B10757_Nevada</th>\n",
       "      <th>B5773_Brazil</th>\n",
       "      <th>Coahuila_2</th>\n",
       "      <th>Colorado_Springs_1</th>\n",
       "      <th>...</th>\n",
       "      <th>Tucson_18</th>\n",
       "      <th>Tucson_19</th>\n",
       "      <th>Tucson_21</th>\n",
       "      <th>Tucson_22</th>\n",
       "      <th>Tucson_23</th>\n",
       "      <th>Tucson_3</th>\n",
       "      <th>Tucson_5</th>\n",
       "      <th>Tucson_6</th>\n",
       "      <th>Tucson_8</th>\n",
       "      <th>Tucson_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>251</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>501</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>751</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>1001</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome  Start  2566_Venezuela  3796_Venezuela  4545-MICE_Venezuela  \\\n",
       "0  CP075068.1      1               1               1                    1   \n",
       "1  CP075068.1    251               1               1                    1   \n",
       "2  CP075068.1    501               1               1                    1   \n",
       "3  CP075068.1    751               1               1                    1   \n",
       "4  CP075068.1   1001               1               1                    1   \n",
       "\n",
       "   730334_Guatemala  B10757_Nevada  B5773_Brazil  Coahuila_2  \\\n",
       "0                 1              1             1           1   \n",
       "1                 1              1             1           1   \n",
       "2                 1              1             1           1   \n",
       "3                 1              1             1           1   \n",
       "4                 1              1             1           1   \n",
       "\n",
       "   Colorado_Springs_1  ...  Tucson_18  Tucson_19  Tucson_21  Tucson_22  \\\n",
       "0                   1  ...          1          1          1          1   \n",
       "1                   1  ...          1          1          1          1   \n",
       "2                   1  ...          1          1          1          1   \n",
       "3                   1  ...          1          1          1          1   \n",
       "4                   1  ...          1          1          1          1   \n",
       "\n",
       "   Tucson_23  Tucson_3  Tucson_5  Tucson_6  Tucson_8  Tucson_9  \n",
       "0          1         1         1         1         1         1  \n",
       "1          1         1         1         1         1         1  \n",
       "2          1         1         1         1         1         1  \n",
       "3          1         1         1         1         1         1  \n",
       "4          1         1         1         1         1         1  \n",
       "\n",
       "[5 rows x 43 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "plain-glasgow",
   "metadata": {},
   "source": [
    "### Add gene info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "applied-shepherd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>direction</th>\n",
       "      <th>annotation</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>gene</td>\n",
       "      <td>57375</td>\n",
       "      <td>57980</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=gene-D8B26_000001;Name=D8B26_000001;gbkey=G...</td>\n",
       "      <td>D8B26_000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>gene</td>\n",
       "      <td>58131</td>\n",
       "      <td>58768</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=gene-D8B26_000002;Name=D8B26_000002;gbkey=G...</td>\n",
       "      <td>D8B26_000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>gene</td>\n",
       "      <td>58628</td>\n",
       "      <td>59933</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=gene-D8B26_000003;Name=D8B26_000003;gbkey=G...</td>\n",
       "      <td>D8B26_000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>gene</td>\n",
       "      <td>60054</td>\n",
       "      <td>62173</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=gene-D8B26_000004;Name=D8B26_000004;gbkey=G...</td>\n",
       "      <td>D8B26_000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>gene</td>\n",
       "      <td>126638</td>\n",
       "      <td>126720</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=gene-D8B26_000005;Name=D8B26_000005;gbkey=G...</td>\n",
       "      <td>D8B26_000005</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom  type   start    stop direction  \\\n",
       "0  CP075068.1  gene   57375   57980         -   \n",
       "1  CP075068.1  gene   58131   58768         -   \n",
       "2  CP075068.1  gene   58628   59933         -   \n",
       "3  CP075068.1  gene   60054   62173         -   \n",
       "4  CP075068.1  gene  126638  126720         +   \n",
       "\n",
       "                                          annotation          name  \n",
       "0  ID=gene-D8B26_000001;Name=D8B26_000001;gbkey=G...  D8B26_000001  \n",
       "1  ID=gene-D8B26_000002;Name=D8B26_000002;gbkey=G...  D8B26_000002  \n",
       "2  ID=gene-D8B26_000003;Name=D8B26_000003;gbkey=G...  D8B26_000003  \n",
       "3  ID=gene-D8B26_000004;Name=D8B26_000004;gbkey=G...  D8B26_000004  \n",
       "4  ID=gene-D8B26_000005;Name=D8B26_000005;gbkey=G...  D8B26_000005  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_path = '../ref/genbank_files/CpSilv.genes_only.gff3'\n",
    "gff = pd.read_csv(gff_path, sep='\\t', header=None)[[0,2,3,4,6,8]]\n",
    "gff.columns = ['chrom', 'type', 'start', 'stop', 'direction', 'annotation']\n",
    "gff['name'] = gff['annotation'].str.split(';',expand=True)[0].str.replace('ID=gene-','')\n",
    "\n",
    "gff = gff[gff['chrom'].isin(['CP075068.1','CP075069.1','CP075070.1','CP075071.1','CP075072.1'])]\n",
    "\n",
    "gff.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "exciting-domain",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CP075068.1\n",
      "CP075069.1\n",
      "CP075070.1\n",
      "CP075071.1\n",
      "CP075072.1\n"
     ]
    }
   ],
   "source": [
    "window_size=250\n",
    "window_genes = []\n",
    "\n",
    "for chrom in gff['chrom'].unique():\n",
    "    \n",
    "    print(chrom)\n",
    "    \n",
    "    temp_genes = gff[gff['chrom']==chrom]\n",
    "    temp_cnv = df[df['Chromosome']==chrom]\n",
    "        \n",
    "    for gene in temp_genes.iterrows():\n",
    "\n",
    "        gene_start,gene_stop = gene[1].start, gene[1].stop\n",
    "            \n",
    "        for window_start in temp_cnv['Start']:\n",
    "            if window_start > gene_stop:\n",
    "                break\n",
    "            \n",
    "            window_stop=window_start+window_size                \n",
    "            overlap = range(max(window_start, gene_start), min(window_stop, gene_stop)+1)\n",
    "\n",
    "            if len(overlap)<50:   \n",
    "                 continue\n",
    "                    \n",
    "            else:\n",
    "                window_genes += [[chrom, window_start, gene[1]['name'], gene[1]['start'], gene[1]['stop']]]\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "variable-district",
   "metadata": {},
   "outputs": [],
   "source": [
    "window_gene_df = pd.DataFrame(window_genes, columns=['Chromosome', 'Start', 'gene', 'gene_start', 'gene_stop'])\n",
    "df = window_gene_df.merge(df, on=['Chromosome', 'Start'], how='outer').sort_values(['Chromosome','Start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "worst-anger",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_stop</th>\n",
       "      <th>2566_Venezuela</th>\n",
       "      <th>3796_Venezuela</th>\n",
       "      <th>4545-MICE_Venezuela</th>\n",
       "      <th>730334_Guatemala</th>\n",
       "      <th>B10757_Nevada</th>\n",
       "      <th>...</th>\n",
       "      <th>Tucson_18</th>\n",
       "      <th>Tucson_19</th>\n",
       "      <th>Tucson_21</th>\n",
       "      <th>Tucson_22</th>\n",
       "      <th>Tucson_23</th>\n",
       "      <th>Tucson_3</th>\n",
       "      <th>Tucson_5</th>\n",
       "      <th>Tucson_6</th>\n",
       "      <th>Tucson_8</th>\n",
       "      <th>Tucson_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>68273</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68274</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>251</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68275</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68276</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>751</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68277</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>1001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome  Start gene  gene_start  gene_stop  2566_Venezuela  \\\n",
       "68273  CP075068.1      1  NaN         NaN        NaN               1   \n",
       "68274  CP075068.1    251  NaN         NaN        NaN               1   \n",
       "68275  CP075068.1    501  NaN         NaN        NaN               1   \n",
       "68276  CP075068.1    751  NaN         NaN        NaN               1   \n",
       "68277  CP075068.1   1001  NaN         NaN        NaN               1   \n",
       "\n",
       "       3796_Venezuela  4545-MICE_Venezuela  730334_Guatemala  B10757_Nevada  \\\n",
       "68273               1                    1                 1              1   \n",
       "68274               1                    1                 1              1   \n",
       "68275               1                    1                 1              1   \n",
       "68276               1                    1                 1              1   \n",
       "68277               1                    1                 1              1   \n",
       "\n",
       "       ...  Tucson_18  Tucson_19  Tucson_21  Tucson_22  Tucson_23  Tucson_3  \\\n",
       "68273  ...          1          1          1          1          1         1   \n",
       "68274  ...          1          1          1          1          1         1   \n",
       "68275  ...          1          1          1          1          1         1   \n",
       "68276  ...          1          1          1          1          1         1   \n",
       "68277  ...          1          1          1          1          1         1   \n",
       "\n",
       "       Tucson_5  Tucson_6  Tucson_8  Tucson_9  \n",
       "68273         1         1         1         1  \n",
       "68274         1         1         1         1  \n",
       "68275         1         1         1         1  \n",
       "68276         1         1         1         1  \n",
       "68277         1         1         1         1  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "professional-brighton",
   "metadata": {},
   "source": [
    "### Calculate Vst in each window"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "small-reporter",
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_dict = json.load(open('../isolate_and_pop_info/pop_dict.json'))\n",
    "    \n",
    "strains = list(pop_dict.keys())\n",
    "sorted_strain_cols = sorted(pop_dict, key=pop_dict.get)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "emerging-elizabeth",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[df.columns[:5].tolist() + sorted_strain_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "fuzzy-publicity",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_stop</th>\n",
       "      <th>Tucson_9</th>\n",
       "      <th>Phoenix_7</th>\n",
       "      <th>Tucson_21</th>\n",
       "      <th>Phoenix_2</th>\n",
       "      <th>Tucson_19</th>\n",
       "      <th>...</th>\n",
       "      <th>4545-MICE_Venezuela</th>\n",
       "      <th>3796_Venezuela</th>\n",
       "      <th>Nuevo_Leon_1</th>\n",
       "      <th>730334_Guatemala</th>\n",
       "      <th>Nuevo_Leon_2</th>\n",
       "      <th>San_Antonio_1</th>\n",
       "      <th>GT002_Texas</th>\n",
       "      <th>Coahuila_2</th>\n",
       "      <th>Sonora_1</th>\n",
       "      <th>B5773_Brazil</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>68273</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68274</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>251</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68275</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68276</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>751</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68277</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>1001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome  Start gene  gene_start  gene_stop  Tucson_9  Phoenix_7  \\\n",
       "68273  CP075068.1      1  NaN         NaN        NaN         1          1   \n",
       "68274  CP075068.1    251  NaN         NaN        NaN         1          1   \n",
       "68275  CP075068.1    501  NaN         NaN        NaN         1          1   \n",
       "68276  CP075068.1    751  NaN         NaN        NaN         1          1   \n",
       "68277  CP075068.1   1001  NaN         NaN        NaN         1          1   \n",
       "\n",
       "       Tucson_21  Phoenix_2  Tucson_19  ...  4545-MICE_Venezuela  \\\n",
       "68273          1          1          1  ...                    1   \n",
       "68274          1          1          1  ...                    1   \n",
       "68275          1          1          1  ...                    1   \n",
       "68276          1          1          1  ...                    1   \n",
       "68277          1          1          1  ...                    1   \n",
       "\n",
       "       3796_Venezuela  Nuevo_Leon_1  730334_Guatemala  Nuevo_Leon_2  \\\n",
       "68273               1             1                 1             1   \n",
       "68274               1             1                 1             1   \n",
       "68275               1             1                 1             1   \n",
       "68276               1             1                 1             1   \n",
       "68277               1             1                 1             1   \n",
       "\n",
       "       San_Antonio_1  GT002_Texas  Coahuila_2  Sonora_1  B5773_Brazil  \n",
       "68273              2            1           1         1             1  \n",
       "68274              2            1           1         1             1  \n",
       "68275              2            1           1         1             1  \n",
       "68276              2            1           1         1             1  \n",
       "68277              2            1           1         1             1  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "adopted-vermont",
   "metadata": {},
   "outputs": [],
   "source": [
    "#for each population, calculate CN mean and variance within that population\n",
    "pops = ['AZ','TXMXSA','CB']\n",
    "\n",
    "var_df = df[list(df.columns[:3])]\n",
    "\n",
    "for pop in pops:\n",
    "    pop_strains = [s for s in strains if pop_dict[s]==pop]\n",
    "    \n",
    "    var_df.loc[:,'CN_MEAN_{}'.format(pop)] = df[pop_strains].apply(np.mean, axis=1)\n",
    "    var_df.loc[:,'VAR_{}'.format(pop)] = df[pop_strains].apply(np.var, axis=1)\n",
    "    var_df.loc[:,'N_{}'.format(pop)] = len(pop_strains)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "electrical-davis",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculate VST\n",
    "def calc_vst_3pops(v_pop1, v_pop2, v_pop3, n_pop1, n_pop2, n_pop3, v_total):\n",
    "    \n",
    "    vn = ((v_pop1*n_pop1)+(v_pop2*n_pop2)+(v_pop3*n_pop3))/(n_pop1+n_pop2+n_pop3)   \n",
    "    vst = (v_total-vn)/v_total\n",
    "    return vst\n",
    "\n",
    "pop1, pop2, pop3 = pops\n",
    "VAR_total = df[strains].apply(np.var, axis=1)\n",
    "var_df.loc[:, 'VST_{}_{}_{}'.format(pop1, pop2, pop3)] = calc_vst_3pops(var_df['VAR_'+pop1], var_df['VAR_'+pop2],\n",
    "                                                                     var_df['VAR_'+pop3],var_df['N_'+pop1], \n",
    "                                                                     var_df['N_'+pop2], var_df['N_'+pop3], VAR_total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "ceramic-valley",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = var_df[['Chromosome', 'Start','VST_AZ_TXMXSA_CB']].merge(df, \n",
    "         on=['Chromosome', 'Start'], how='outer').sort_values(['Chromosome','Start'])\n",
    "df = df.drop(columns=['gene_start', 'gene_stop'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "linear-dublin",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('../tables/TableS3B_VST.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "alike-costume",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py3",
   "language": "python",
   "name": "py3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
