{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "running-palestinian",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import dendropy\n",
    "import csv\n",
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "import json\n",
    "from collections import Counter\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "gothic-madagascar",
   "metadata": {},
   "source": [
    "### Split AZ population into strains with the chromosome III locus present or with the locus absent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "strange-chile",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "29"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#add pop info\n",
    "pop_dict = json.load(open('../isolate_and_pop_info/pop_dict.json'))\n",
    "for strain, pops in pop_dict.items():\n",
    "    pop_dict[strain] = pops\n",
    "    \n",
    "AZ_strains = [i for i, pop in pop_dict.items() if pop=='AZ']\n",
    "len(AZ_strains)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "corrected-religion",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('AZ_strains.txt', 'w') as f:\n",
    "\n",
    "    for strain in AZ_strains:\n",
    "        f.write(strain+'\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "amazing-patio",
   "metadata": {},
   "outputs": [],
   "source": [
    "AZ_with = [\"Phoenix_1\",\"Phoenix_3\",\"Phoenix_6\",\"Tucson_14\",\"Phoenix_4\",\"Tucson_12\",\"Tucson_1\",\"Tucson_18\",\"Tucson_17\",\"Tucson_10\",]\n",
    "AZ_without =  [\"GT162_USA\",\"Colorado_Springs_1\",\"Tucson_9\",\"Phoenix_7\",\"Tucson_21\",\"Phoenix_2\",\"Tucson_19\",\"Phoenix_9\",\n",
    "              \"B10757_Nevada\",\"Tucson_23\",\"Phoenix_5\",\"Phoenix_8\",\"Tucson_15\",\"Tucson_3\",\"Tucson_6\",\"Tucson_5\",\"Tucson_8\",\n",
    "              \"Tucson_13\",\"Tucson_22\",]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "with open('AZ_strains_locus_present.txt', 'w') as f:\n",
    "\n",
    "    for strain in AZ_with:\n",
    "        f.write(strain+'\\n')\n",
    "        \n",
    "with open('AZ_strains_locus_absent.txt', 'w') as f:\n",
    "\n",
    "    for strain in AZ_without:\n",
    "        f.write(strain+'\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "metropolitan-momentum",
   "metadata": {},
   "source": [
    "### Calculate Tajima's D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "accurate-forest",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--vcf ../isolate_and_pop_info/all_strains_variant_sites.renamed.vcf\n",
      "\t--keep AZ_strains_locus_present.txt\n",
      "\t--max-missing 1\n",
      "\t--out AZ_strains_locus_present_250bp_window\n",
      "\t--TajimaD 250\n",
      "\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=BC,Number=4,Type=Integer,Description=\"Count of As, Cs, Gs, Ts at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=BC,Number=4,Type=Integer,Description=\"Count of As, Cs, Gs, Ts at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=BC,Number=4,Type=Integer,Description=\"Count of As, Cs, Gs, Ts at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=QP,Number=4,Type=Integer,Description=\"Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=QP,Number=4,Type=Integer,Description=\"Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=QP,Number=4,Type=Integer,Description=\"Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes, for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes, for each ALT allele, in the same order as listed\">\n",
      "Keeping individuals in 'keep' list\n",
      "After filtering, kept 10 out of 61 Individuals\n",
      "Outputting Tajima's D Statistic...\n",
      "\tTajimaD: Only using bialleleic sites.\n",
      "After filtering, kept 404632 out of a possible 1012365 Sites\n",
      "Run Time = 16.00 seconds\n"
     ]
    }
   ],
   "source": [
    "##Tajima's D for strains with locus present\n",
    "\n",
    "vcf = '../isolate_and_pop_info/all_strains_variant_sites.renamed.vcf'\n",
    "\n",
    "!vcftools --vcf {vcf} --TajimaD 250 --max-missing 1.0 --keep AZ_strains_locus_present.txt --out AZ_strains_locus_present_250bp_window\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "maritime-rachel",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--vcf ../isolate_and_pop_info/all_strains_variant_sites.renamed.vcf\n",
      "\t--keep AZ_strains_locus_absent.txt\n",
      "\t--max-missing 1\n",
      "\t--out AZ_strains_locus_absent_250bp_window\n",
      "\t--TajimaD 250\n",
      "\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=BC,Number=4,Type=Integer,Description=\"Count of As, Cs, Gs, Ts at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=BC,Number=4,Type=Integer,Description=\"Count of As, Cs, Gs, Ts at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=BC,Number=4,Type=Integer,Description=\"Count of As, Cs, Gs, Ts at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=QP,Number=4,Type=Integer,Description=\"Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=QP,Number=4,Type=Integer,Description=\"Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=QP,Number=4,Type=Integer,Description=\"Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes, for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes, for each ALT allele, in the same order as listed\">\n",
      "Keeping individuals in 'keep' list\n",
      "After filtering, kept 19 out of 61 Individuals\n",
      "Outputting Tajima's D Statistic...\n",
      "\tTajimaD: Only using bialleleic sites.\n",
      "After filtering, kept 360743 out of a possible 1012365 Sites\n",
      "Run Time = 16.00 seconds\n"
     ]
    }
   ],
   "source": [
    "##Tajima's D for strains with locus absent\n",
    "\n",
    "vcf = '../isolate_and_pop_info/all_strains_variant_sites.renamed.vcf'\n",
    "\n",
    "!vcftools --vcf {vcf} --TajimaD 250 --max-missing 1.0 --keep AZ_strains_locus_absent.txt --out AZ_strains_locus_absent_250bp_window\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "natural-department",
   "metadata": {},
   "outputs": [],
   "source": [
    "td_present = pd.read_csv('AZ_strains_locus_present_250bp_window.Tajima.D', sep='\\t').drop(columns='N_SNPS')\n",
    "td_absent = pd.read_csv('AZ_strains_locus_absent_250bp_window.Tajima.D', sep='\\t').drop(columns='N_SNPS')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "exterior-image",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>BIN_START</th>\n",
       "      <th>TajimaD_present</th>\n",
       "      <th>TajimaD_absent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>501</td>\n",
       "      <td>-0.591550</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>751</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.825509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>1001</td>\n",
       "      <td>-0.768572</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>1251</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>1501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.090600</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHROM  BIN_START  TajimaD_present  TajimaD_absent\n",
       "0  CP075068.1        501        -0.591550             NaN\n",
       "1  CP075068.1        751              NaN       -0.825509\n",
       "2  CP075068.1       1001        -0.768572             NaN\n",
       "3  CP075068.1       1251              NaN             NaN\n",
       "4  CP075068.1       1501              NaN       -1.090600"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "td = td_present.merge(td_absent, on=['CHROM','BIN_START'], suffixes=['_present','_absent'], how='outer')\n",
    "td['BIN_START'] = td['BIN_START']+1\n",
    "\n",
    "td.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "protected-probability",
   "metadata": {},
   "source": [
    "### Calculate FST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "pending-photograph",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--vcf ../isolate_and_pop_info/all_strains_variant_sites.renamed.vcf\n",
      "\t--fst-window-size 250\n",
      "\t--fst-window-step 250\n",
      "\t--weir-fst-pop AZ_strains_locus_present.txt\n",
      "\t--weir-fst-pop AZ_strains_locus_absent.txt\n",
      "\t--keep AZ_strains_locus_present.txt\n",
      "\t--keep AZ_strains_locus_absent.txt\n",
      "\t--max-missing 1\n",
      "\t--out AZ_strains_250bp_window\n",
      "\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=BC,Number=4,Type=Integer,Description=\"Count of As, Cs, Gs, Ts at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=BC,Number=4,Type=Integer,Description=\"Count of As, Cs, Gs, Ts at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=BC,Number=4,Type=Integer,Description=\"Count of As, Cs, Gs, Ts at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=QP,Number=4,Type=Integer,Description=\"Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=QP,Number=4,Type=Integer,Description=\"Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=QP,Number=4,Type=Integer,Description=\"Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes, for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes, for each ALT allele, in the same order as listed\">\n",
      "Keeping individuals in 'keep' list\n",
      "After filtering, kept 29 out of 61 Individuals\n",
      "Outputting Windowed Weir and Cockerham Fst estimates.\n",
      "Weir and Cockerham mean Fst estimate: -0.0040478\n",
      "Weir and Cockerham weighted Fst estimate: 0.0019009\n",
      "After filtering, kept 339399 out of a possible 1012365 Sites\n",
      "Run Time = 17.00 seconds\n"
     ]
    }
   ],
   "source": [
    "vcf = '../isolate_and_pop_info/all_strains_variant_sites.renamed.vcf'\n",
    "\n",
    "!vcftools --vcf {vcf} --fst-window-size 250 --fst-window-step 250 --max-missing 1.0 --weir-fst-pop  AZ_strains_locus_present.txt --weir-fst-pop  AZ_strains_locus_absent.txt --out AZ_strains_250bp_window"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "assured-double",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>BIN_START</th>\n",
       "      <th>WEIGHTED_FST</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>751</td>\n",
       "      <td>-0.038251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>1501</td>\n",
       "      <td>-0.038251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>2251</td>\n",
       "      <td>0.068627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>4501</td>\n",
       "      <td>-0.038251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>66251</td>\n",
       "      <td>-0.019324</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHROM  BIN_START  WEIGHTED_FST\n",
       "0  CP075068.1        751     -0.038251\n",
       "1  CP075068.1       1501     -0.038251\n",
       "2  CP075068.1       2251      0.068627\n",
       "3  CP075068.1       4501     -0.038251\n",
       "4  CP075068.1      66251     -0.019324"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fst = pd.read_csv('AZ_strains_250bp_window.windowed.weir.fst', sep='\\t').drop(columns=['BIN_END', 'MEAN_FST', 'N_VARIANTS'])\n",
    "fst.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "crude-subject",
   "metadata": {},
   "source": [
    "### Calculate pi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "prostate-attitude",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--vcf ../isolate_and_pop_info/all_strains_variant_sites.renamed.vcf\n",
      "\t--keep AZ_strains_locus_present.txt\n",
      "\t--max-missing 1\n",
      "\t--out AZ_strains_locus_present_250bp_window\n",
      "\t--window-pi 250\n",
      "\t--window-pi-step 250\n",
      "\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=BC,Number=4,Type=Integer,Description=\"Count of As, Cs, Gs, Ts at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=BC,Number=4,Type=Integer,Description=\"Count of As, Cs, Gs, Ts at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=BC,Number=4,Type=Integer,Description=\"Count of As, Cs, Gs, Ts at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=QP,Number=4,Type=Integer,Description=\"Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=QP,Number=4,Type=Integer,Description=\"Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=QP,Number=4,Type=Integer,Description=\"Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes, for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes, for each ALT allele, in the same order as listed\">\n",
      "Keeping individuals in 'keep' list\n",
      "After filtering, kept 10 out of 61 Individuals\n",
      "Outputting Windowed Nucleotide Diversity Statistics...\n",
      "After filtering, kept 404632 out of a possible 1012365 Sites\n",
      "Run Time = 16.00 seconds\n"
     ]
    }
   ],
   "source": [
    "vcf = '../isolate_and_pop_info/all_strains_variant_sites.renamed.vcf'\n",
    "\n",
    "!vcftools --vcf {vcf} --window-pi 250 --window-pi-step 250 --max-missing 1.0 --keep AZ_strains_locus_present.txt --out AZ_strains_locus_present_250bp_window\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "promotional-efficiency",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.17\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--vcf ../isolate_and_pop_info/all_strains_variant_sites.renamed.vcf\n",
      "\t--keep AZ_strains_locus_absent.txt\n",
      "\t--max-missing 1\n",
      "\t--out AZ_strains_locus_absent_250bp_window\n",
      "\t--window-pi 250\n",
      "\t--window-pi-step 250\n",
      "\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=BC,Number=4,Type=Integer,Description=\"Count of As, Cs, Gs, Ts at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=BC,Number=4,Type=Integer,Description=\"Count of As, Cs, Gs, Ts at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=BC,Number=4,Type=Integer,Description=\"Count of As, Cs, Gs, Ts at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=QP,Number=4,Type=Integer,Description=\"Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=QP,Number=4,Type=Integer,Description=\"Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=QP,Number=4,Type=Integer,Description=\"Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes, for each ALT allele, in the same order as listed\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes, for each ALT allele, in the same order as listed\">\n",
      "Keeping individuals in 'keep' list\n",
      "After filtering, kept 19 out of 61 Individuals\n",
      "Outputting Windowed Nucleotide Diversity Statistics...\n",
      "After filtering, kept 360743 out of a possible 1012365 Sites\n",
      "Run Time = 16.00 seconds\n"
     ]
    }
   ],
   "source": [
    "vcf = '../isolate_and_pop_info/all_strains_variant_sites.renamed.vcf'\n",
    "\n",
    "!vcftools --vcf {vcf} --window-pi 250 --window-pi-step 250 --max-missing 1.0 --keep AZ_strains_locus_absent.txt --out AZ_strains_locus_absent_250bp_window\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "mathematical-munich",
   "metadata": {},
   "outputs": [],
   "source": [
    "pi_present = pd.read_csv('AZ_strains_locus_present_250bp_window.windowed.pi', sep='\\t').drop(columns=['BIN_END', 'N_VARIANTS'])\n",
    "pi_absent = pd.read_csv('AZ_strains_locus_absent_250bp_window.windowed.pi', sep='\\t').drop(columns=['BIN_END', 'N_VARIANTS'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "unavailable-africa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>BIN_START</th>\n",
       "      <th>PI_present</th>\n",
       "      <th>PI_absent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>501</td>\n",
       "      <td>0.000758</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>1001</td>\n",
       "      <td>0.001516</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>2251</td>\n",
       "      <td>0.000758</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>5251</td>\n",
       "      <td>0.000758</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>5501</td>\n",
       "      <td>0.000758</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHROM  BIN_START  PI_present  PI_absent\n",
       "0  CP075068.1        501    0.000758        NaN\n",
       "1  CP075068.1       1001    0.001516        NaN\n",
       "2  CP075068.1       2251    0.000758        NaN\n",
       "3  CP075068.1       5251    0.000758        NaN\n",
       "4  CP075068.1       5501    0.000758        NaN"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pi = pi_present.merge(pi_absent, on=['CHROM','BIN_START'], suffixes=['_present','_absent'], how='outer')\n",
    "pi.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "valued-context",
   "metadata": {},
   "source": [
    "### Merge all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "falling-enhancement",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>BIN_START</th>\n",
       "      <th>TajimaD_present</th>\n",
       "      <th>TajimaD_absent</th>\n",
       "      <th>PI_present</th>\n",
       "      <th>PI_absent</th>\n",
       "      <th>WEIGHTED_FST</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>501</td>\n",
       "      <td>-0.591550</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000758</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>751</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.825509</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000410</td>\n",
       "      <td>-0.038251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>1001</td>\n",
       "      <td>-0.768572</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001516</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>1251</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP075068.1</td>\n",
       "      <td>1501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.090600</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000819</td>\n",
       "      <td>-0.038251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111169</th>\n",
       "      <td>CP075072.1</td>\n",
       "      <td>1488001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111170</th>\n",
       "      <td>CP075072.1</td>\n",
       "      <td>1488251</td>\n",
       "      <td>-0.591550</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000758</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111171</th>\n",
       "      <td>CP075072.1</td>\n",
       "      <td>1488501</td>\n",
       "      <td>0.351953</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001347</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111172</th>\n",
       "      <td>CP075072.1</td>\n",
       "      <td>1488751</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.694900</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001957</td>\n",
       "      <td>0.007223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111173</th>\n",
       "      <td>CP075072.1</td>\n",
       "      <td>1489001</td>\n",
       "      <td>-0.591550</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000758</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.068627</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111174 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             CHROM  BIN_START  TajimaD_present  TajimaD_absent  PI_present  \\\n",
       "0       CP075068.1        501        -0.591550             NaN    0.000758   \n",
       "1       CP075068.1        751              NaN       -0.825509         NaN   \n",
       "2       CP075068.1       1001        -0.768572             NaN    0.001516   \n",
       "3       CP075068.1       1251              NaN             NaN         NaN   \n",
       "4       CP075068.1       1501              NaN       -1.090600         NaN   \n",
       "...            ...        ...              ...             ...         ...   \n",
       "111169  CP075072.1    1488001              NaN             NaN         NaN   \n",
       "111170  CP075072.1    1488251        -0.591550             NaN    0.000758   \n",
       "111171  CP075072.1    1488501         0.351953             NaN    0.001347   \n",
       "111172  CP075072.1    1488751              NaN       -0.694900         NaN   \n",
       "111173  CP075072.1    1489001        -0.591550             NaN    0.000758   \n",
       "\n",
       "        PI_absent  WEIGHTED_FST  \n",
       "0             NaN           NaN  \n",
       "1        0.000410     -0.038251  \n",
       "2             NaN           NaN  \n",
       "3             NaN           NaN  \n",
       "4        0.000819     -0.038251  \n",
       "...           ...           ...  \n",
       "111169        NaN           NaN  \n",
       "111170        NaN           NaN  \n",
       "111171        NaN           NaN  \n",
       "111172   0.001957      0.007223  \n",
       "111173        NaN      0.068627  \n",
       "\n",
       "[111174 rows x 7 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged = td.merge(pi, on=['CHROM', 'BIN_START'], how='outer').merge(fst, on=['CHROM', 'BIN_START'], how='outer')\n",
    "merged = merged.sort_values(['CHROM', 'BIN_START'])\n",
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "suffering-patent",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.to_csv('../tables/TableS5_present_absent_stats.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "confused-pontiac",
   "metadata": {},
   "outputs": [],
   "source": [
    "start = 1237751\n",
    "stop = 1428001\n",
    "\n",
    "merged_upstream = merged[(merged['CHROM']=='CP075070.1') & (merged['BIN_START'].isin(range(start-5000, start+250)))]\n",
    "merged_downstream = merged[(merged['CHROM']=='CP075070.1') & (merged['BIN_START'].isin(range(stop-250, stop+5000)))]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "particular-window",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(400)\n",
    "\n",
    "p_dict = {}\n",
    "\n",
    "for stat in ['PI_present','PI_absent', 'WEIGHTED_FST']:\n",
    "    \n",
    "    max_val = merged_upstream[stat].max()\n",
    "    sample_counts = []\n",
    "\n",
    "    for _ in range(10000):\n",
    "    \n",
    "        sample_index = merged.sample().index[0]\n",
    "        sample = merged.loc[sample_index:sample_index+(5000/250)]\n",
    "        sample_over_threshold = sample[sample[stat] >= float(max_val)].shape[0]    \n",
    "        sample_counts += [sample_over_threshold]\n",
    "\n",
    "\n",
    "    p = len([i for i in sample_counts if i >= 1])/len(sample_counts)\n",
    "    p_dict[stat+'_upstream'] = [max_val, p]\n",
    "\n",
    "\n",
    "for stat in ['TajimaD_present', 'TajimaD_absent']:\n",
    "    \n",
    "    min_val = merged_upstream[stat].min()\n",
    "    sample_counts = []\n",
    "\n",
    "    for _ in range(10000):\n",
    "    \n",
    "        sample_index = merged.sample().index[0]\n",
    "        sample = merged.loc[sample_index:sample_index+(5000/250)]\n",
    "        sample_over_threshold = sample[sample[stat] <= float(min_val)].shape[0]    \n",
    "        sample_counts += [sample_over_threshold]\n",
    "\n",
    "\n",
    "    p = len([i for i in sample_counts if i >= 1])/len(sample_counts)\n",
    "    p_dict[stat+'_upstream'] = [min_val, p]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "affecting-christopher",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(500)\n",
    "\n",
    "for stat in ['PI_present','PI_absent', 'WEIGHTED_FST']:\n",
    "    \n",
    "    max_val = merged_downstream[stat].max()\n",
    "    sample_counts = []\n",
    "\n",
    "    for _ in range(10000):\n",
    "    \n",
    "        sample_index = merged.sample().index[0]\n",
    "        sample = merged.loc[sample_index:sample_index+(5000/250)]\n",
    "        sample_over_threshold = sample[sample[stat] >= float(max_val)].shape[0]    \n",
    "        sample_counts += [sample_over_threshold]\n",
    "\n",
    "\n",
    "    p = len([i for i in sample_counts if i >= 1])/len(sample_counts)\n",
    "    p_dict[stat+'_downstream'] = [max_val, p]\n",
    "\n",
    "\n",
    "for stat in ['TajimaD_present', 'TajimaD_absent']:\n",
    "    \n",
    "    min_val = merged_downstream[stat].min()\n",
    "    sample_counts = []\n",
    "\n",
    "    for _ in range(10000):\n",
    "    \n",
    "        sample_index = merged.sample().index[0]\n",
    "        sample = merged.loc[sample_index:sample_index+(5000/250)]\n",
    "        sample_over_threshold = sample[sample[stat] <= float(min_val)].shape[0]    \n",
    "        sample_counts += [sample_over_threshold]\n",
    "\n",
    "\n",
    "    p = len([i for i in sample_counts if i >= 1])/len(sample_counts)\n",
    "    p_dict[stat+'_downstream'] = [min_val, p]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "laden-mixture",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'PI_present_upstream': [0.00614737, 0.2926],\n",
       " 'PI_absent_upstream': [0.00505263, 0.4055],\n",
       " 'WEIGHTED_FST_upstream': [0.0420787, 0.7775],\n",
       " 'TajimaD_present_upstream': [-1.003, 0.0675],\n",
       " 'TajimaD_absent_upstream': [-1.25802, 0.5038],\n",
       " 'PI_present_downstream': [0.0048, 0.5239],\n",
       " 'PI_absent_downstream': [0.00669132, 0.198],\n",
       " 'WEIGHTED_FST_downstream': [0.0686275, 0.7124],\n",
       " 'TajimaD_present_downstream': [-0.948945, 0.1454],\n",
       " 'TajimaD_absent_downstream': [-1.37697, 0.2305]}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "impressed-gender",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_df = pd.DataFrame.from_dict(p_dict, orient='index')\n",
    "p_df.columns = ['threshold_tested', 'p_val']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "green-button",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>threshold_tested</th>\n",
       "      <th>p_val</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PI_present_upstream</th>\n",
       "      <td>0.006147</td>\n",
       "      <td>0.2926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI_absent_upstream</th>\n",
       "      <td>0.005053</td>\n",
       "      <td>0.4055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WEIGHTED_FST_upstream</th>\n",
       "      <td>0.042079</td>\n",
       "      <td>0.7775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TajimaD_present_upstream</th>\n",
       "      <td>-1.003000</td>\n",
       "      <td>0.0675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TajimaD_absent_upstream</th>\n",
       "      <td>-1.258020</td>\n",
       "      <td>0.5038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI_present_downstream</th>\n",
       "      <td>0.004800</td>\n",
       "      <td>0.5239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI_absent_downstream</th>\n",
       "      <td>0.006691</td>\n",
       "      <td>0.1980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WEIGHTED_FST_downstream</th>\n",
       "      <td>0.068627</td>\n",
       "      <td>0.7124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TajimaD_present_downstream</th>\n",
       "      <td>-0.948945</td>\n",
       "      <td>0.1454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TajimaD_absent_downstream</th>\n",
       "      <td>-1.376970</td>\n",
       "      <td>0.2305</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            threshold_tested   p_val\n",
       "PI_present_upstream                 0.006147  0.2926\n",
       "PI_absent_upstream                  0.005053  0.4055\n",
       "WEIGHTED_FST_upstream               0.042079  0.7775\n",
       "TajimaD_present_upstream           -1.003000  0.0675\n",
       "TajimaD_absent_upstream            -1.258020  0.5038\n",
       "PI_present_downstream               0.004800  0.5239\n",
       "PI_absent_downstream                0.006691  0.1980\n",
       "WEIGHTED_FST_downstream             0.068627  0.7124\n",
       "TajimaD_present_downstream         -0.948945  0.1454\n",
       "TajimaD_absent_downstream          -1.376970  0.2305"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "closed-dimension",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_df.to_csv('../tables/TableS5_p_vals.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py3",
   "language": "python",
   "name": "py3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
